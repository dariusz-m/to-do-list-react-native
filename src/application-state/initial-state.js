export const INITIAL_STATE = {
    todo: {
        tasks: [],
        editingTask: null,
        newTask: '',
        errorMessage: '',
    },
    auth: {
        errorMessage: '',
    },

};
