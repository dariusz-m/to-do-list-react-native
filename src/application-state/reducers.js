import { combineReducers } from 'redux';

import toDoReducer from '../tasks/reducer';
import authReducer from '../authentication/reducer';

export default combineReducers({
    todo: toDoReducer,
    auth: authReducer,
});
