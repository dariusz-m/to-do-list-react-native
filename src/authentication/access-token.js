import { AsyncStorage } from 'react-native';

const ACCESS_TOKEN_KEY = 'access-token';

/**
 * Store user access token
 * @param {string} token - Access token
 */
export const storeAccessToken = async (token) => {
    try {
        await AsyncStorage.setItem(ACCESS_TOKEN_KEY, token);
    } catch (error) {}
};

/**
 * Get access token from storage
 * @returns {string | null}
 */
export const getAccessToken = async () => {
    try {
        return await AsyncStorage.getItem(ACCESS_TOKEN_KEY);
    } catch (error) {
        return null;
    }
};

/**
 * Remove access token.
 */
export const removeAccessToken = async () => {
    try {
        await AsyncStorage.removeItem(ACCESS_TOKEN_KEY);
    } catch (error) {}
};
