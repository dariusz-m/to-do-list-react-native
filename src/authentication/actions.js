export const LOGIN = 'LOGIN';
export const LOGIN_FAILED = 'LOGIN_FAILED';
export const USER_HAS_BEEN_LOGGED_IN = 'USER_HAS_BEEN_LOGGED_IN';

/**
 * Login user.
 * @param {string} username - Username
 * @param {string} password - User password
 * @returns {{type: string, payload: {email: string, password: string}}}
 */
export const login = (username, password) => ({
    type: LOGIN,
    payload: {
        username,
        password,
    },
});


/**
 * Login failed
 * @param {string} errorMessage - Error message
 * @returns {{type: string, errorMessage: string}}
 */
export const loginFailed = errorMessage => ({
    type: LOGIN_FAILED,
    payload: {
        errorMessage,
    },
});

/**
 * User has been logged in.
 * @returns {{type: string}}
 */
export const userHasBeenLoggedIn = () => ({
    type: USER_HAS_BEEN_LOGGED_IN,
});
