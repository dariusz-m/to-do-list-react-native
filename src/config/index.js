/* eslint-disable no-undef */
import devConfig from './config.development';
import testConfig from './config.test';

const env = process.env.NODE_ENV || 'development';

const config = {
    development: devConfig,
    test: testConfig,
};

export default config[env];
