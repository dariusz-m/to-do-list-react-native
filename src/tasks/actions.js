export const GET_TASKS = 'GET_TASKS';
export const TASKS_HAVE_BEEN_LOADED = 'TASKS_HAVE_BEEN_LOADED';
export const ADD_TASK_TO_DO = 'ADD_TASK_TO_DO';
export const CHANGE_THE_DESCRIPTION_OF_THE_NEW_TASK = 'CHANGE_THE_DESCRIPTION_OF_THE_NEW_TASK';
export const TASK_HAS_BEEN_ADDED = 'TASK_HAS_BEEN_ADDED';
export const DELETE_TASK = 'DELETE_TASK';
export const TASK_HAS_BEEN_DELETED = 'TASK_HAS_BEEN_DELETED';
export const TASK_HAS_BEEN_UPDATED = 'TASK_HAS_BEEN_UPDATED';
export const UPDATE_TASK = 'UPDATE_TASK';
export const EDIT_TASK = 'EDIT_TASK';
export const CANCEL_EDITING_TASK = 'CANCEL_EDITING_TASK';
export const CHANGE_EDITING_TASK = 'CHANGE_EDITING_TASK';
export const TASKS_REQUEST_FAILED = 'TASKS_REQUEST_FAILED';

/**
 * Get all of the tasks.
 * @returns {{type: string}}
 */
export const getTasks = () => ({
    type: GET_TASKS,
});

/**
 * Tasks request failed.
 * @param {string} errorMessage - Error message
 * @returns {{type: string, payload: {errorMessage: string}}}
 */
export const tasksRequestFailed = errorMessage => ({
    type: TASKS_REQUEST_FAILED,
    payload: {
        errorMessage,
    },
});

/**
 * Tasks have been loaded
 * @param {Array<string>} tasks - Tasks
 * @returns {{type: string, payload: {tasks: Array<string>}}}
 */
export const tasksHaveBeenLoaded = tasks => ({
    type: TASKS_HAVE_BEEN_LOADED,
    payload: {
        tasks,
    },
});


/**
 * Add new task to do
 * @param {string} task - New task
 * @returns {{type: string, payload: {tasks: string}}}
 */
export const addTaskToDo = task => ({
    type: ADD_TASK_TO_DO,
    payload: {
        task,
    },
});


/**
 * Change the description of the new task.
 * @param {string} newTaskDesc - New description of the task
 * @returns {{type: string, payload: {taskId: number, task: string}}}
 */
export const changeTheDescriptionOfTheNewTask = newTaskDesc => ({
    type: CHANGE_THE_DESCRIPTION_OF_THE_NEW_TASK,
    payload: {
        newTaskDesc,
    },
});

/**
 * Task has been added
 * @param {number} taskId - Task id
 * @param {string} task - Task to do
 * @returns {{type: string, payload: {taskId: number, task: string}}}
 */
export const taskHasBeenAdded = (taskId, task) => ({
    type: TASK_HAS_BEEN_ADDED,
    payload: {
        taskId,
        task,
    },
});


/**
 * Delete task
 * @param {number} taskId - Task id
 * @returns {{type: string, payload: {taskId: number}}}
 */
export const deleteTask = taskId => ({
    type: DELETE_TASK,
    payload: {
        taskId,
    },
});

/**
 * Task has been deleted
 * @param {number} taskId - Task id
 * @returns {{type: string, payload: {taskId: number}}}
 */
export const taskHasBeenDeleted = taskId => ({
    type: TASK_HAS_BEEN_DELETED,
    payload: {
        taskId,
    },
});

/**
 * Update task
 * @param {number} taskId - Task id
 * @param {string} newTaskDesc - New task description
 * @returns {{type: string, payload: {taskId: number, newTaskDesc: string}}}
 */
export const updateTask = (taskId, newTaskDesc) => ({
    type: UPDATE_TASK,
    payload: {
        taskId,
        newTaskDesc,
    },
});

/**
 * Task has been updated.
 * @param {number} taskId - Task id
 * @param {string} newTaskDesc - New task description
 * @returns {{type: string, payload: {taskId: number, newTaskDesc: string}}}
 */
export const taskHasBeenUpdated = (taskId, newTaskDesc) => ({
    type: TASK_HAS_BEEN_UPDATED,
    payload: {
        taskId,
        newTaskDesc,
    },
});

/**
 * Edit task
 * @param {number} taskId - Task id
 * @returns {{type: string, payload: {taskId: number}}}
 */
export const editTask = taskId => ({
    type: EDIT_TASK,
    payload: {
        taskId,
    },
});

/**
 * Cancel process of editing task
 * @returns {{type: string}}
 */
export const cancelEditingTask = () => ({
    type: CANCEL_EDITING_TASK,
});

/**
 * Change editing task
 * @param {string} newTaskDesc - New task description
 * @returns {{type: string, payload: {newTaskDesc: string}}}
 */
export const changeEditingTask = newTaskDesc => ({
    type: CHANGE_EDITING_TASK,
    payload: {
        newTaskDesc,
    },
});
