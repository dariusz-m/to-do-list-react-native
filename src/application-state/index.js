import { createStore, applyMiddleware } from 'redux';
import createSagaMiddleware from 'redux-saga';

import reducers from './reducers';
import { INITIAL_STATE } from './initial-state';
import rootSaga from './sagas';

/**
 * Configure store.
 * @param {Object} initialState: Initial state for application state.
 * @returns {Object}
 */
export default function configureStore(initialState = INITIAL_STATE) {
    const sagaMiddleware = createSagaMiddleware();
    const store = createStore(
        reducers,
        initialState,
        applyMiddleware(sagaMiddleware),
    );

    if (module.hot) {
        module.hot.accept('./reducers', () => {
            // eslint-disable-next-line global-require
            const nextRootReducer = require('./reducers');
            store.replaceReducer(nextRootReducer);
        });
    }

    sagaMiddleware.run(rootSaga);

    return store;
}
