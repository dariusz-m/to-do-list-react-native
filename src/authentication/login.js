import React from 'react';
import { Text, View, Button } from 'react-native';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { login } from './actions';
import Input from '../shared-components/input';

/**
 * Login component.
 */
export class Login extends React.Component {
    static propTypes = {
        errorMessage: PropTypes.string.isRequired,
        actions: PropTypes.shape({
            login: PropTypes.func.isRequired,
        }).isRequired,
    };

    state = {
        username: '',
        password: '',
    };

    /**
     * On username change.
     *
     * @param {string} value: Input value
     */
    usernameChange = (value) => {
        this.setState({ username: value });
    };

    /**
     * On password change.
     *
     * @param {string} value: Input value
     */
    passwordChange = (value) => {
        this.setState({ password: value });
    };

    /**
     * Login.
     */
    login = () => {
        this.props.actions.login(this.state.username, this.state.password);
        this.setState({ password: '' });
    };

    /**
     * Render component.
     *
     * @returns {JSXElement}
     */
    render() {
        return (
            <View style={{ textAlign: 'center' }}>
                <Text style={{ fontSize: 40, fontWeight: 'bold' }}>Sign In</Text>
                <Input
                    placeholder={'username...'}
                    onChange={this.usernameChange}
                    label={'Username: '}
                    autoCorrect={true}
                    secureTextEntry={false}
                    value={this.state.username}
                />
                <Input
                    placeholder={'password...'}
                    onChange={this.passwordChange}
                    label={'Password: '}
                    autoCorrect={true}
                    secureTextEntry={true}
                    value={this.state.password}
                />
                <Button onPress={this.login} title={'Sign in'} color={'blue'}/>
                {this.props.errorMessage
                    ? <Text style={{ color: 'red', textAlign: 'center' }}>
                        <Text>Login failed!{'\n'}</Text>
                        <Text>{this.props.errorMessage}</Text>
                    </Text> : null
                }
            </View>
        );
    }
}

/**
 * Map state to props.
 *
 * @param {Object} state: State of application
 *
 * @returns {Object}
 */
function mapStateToProps(state) {
    return {
        errorMessage: state.auth.errorMessage,
    };
}

/**
 * Map state to props.
 *
 * @param {Object} dispatch: Dispatch
 *
 * @returns {Object}
 */
function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators({
            login,
        }, dispatch),
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(Login);
