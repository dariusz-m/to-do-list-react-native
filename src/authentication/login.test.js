import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import ReactTestUtils from 'react-dom/test-utils';

import { Login } from './login';
import config from '../config';
import { INITIAL_STATE } from '../application-state/initial-state';
import { login } from './actions';
import configureStore from '../application-state';
import { getAccessToken, removeAccessToken } from './access-token';


describe('Auth test', () => {
    let store;

    beforeEach(() => {
        fetch.resetMocks();
        store = configureStore();
        removeAccessToken();
    });

    afterAll(() => {
        fetch.resetMocks();
    });

    it('Login component renders correctly without error message', () => {
        const actions = { login: jest.fn() };

        const tree = renderer.create(<Login errorMessage={''} actions={actions}/>).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('Login component renders correctly with error message', () => {
        const actions = { login: jest.fn() };

        const tree = renderer.create(<Login errorMessage={'Bad credentials'} actions={actions}/>).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('After clicking the button[submit] callback(login) is called out once', () => {
        const actions = { login: jest.fn() };
        const component = ReactTestUtils.renderIntoDocument(
            <Login actions={actions} errorMessage={''}/>,
        );
        const username = 'username';
        const password = 'password';

        fillUpUsernameField(component, username);
        fillUpPasswordField(component, password);
        clickSubmitButton(component);

        expectedCallbackOnFormSubmitCalledOnceWith(actions.login, username, password);
    });


    it('Test user can login in and his access token is stored', (done) => {
        expect.hasAssertions();
        const accessToken = 'FDFd3r56hh6h7h7h7h';
        fetch.mockResponse(JSON.stringify({ accessToken }), { status: 200 });

        store.dispatch(login('username', 'password'));

        store.subscribe(() => {
            expect(fetch.mock.calls[0][0]).toEqual(`${config.API_URL}login`);
            expect(getAccessToken()).toEqual(accessToken);
            done();
        });
    });

    it('Test handle error when something went wrong', (done) => {
        expect.hasAssertions();
        const errorMessage = 'Something went wrong, try again.';
        const expectedState = { ...INITIAL_STATE, auth: { ...INITIAL_STATE.auth, errorMessage } };
        fetch.mockReject();

        store.dispatch(login('username', 'password'));

        store.subscribe(() => {
            expect(fetch.mock.calls[0][0]).toEqual(`${config.API_URL}login`);
            expect(store.getState()).toEqual(expectedState);
            expect(getAccessToken()).toBeNull();
            done();
        });
    });

    it('Test handle error on bad credentials', (done) => {
        expect.hasAssertions();
        const errorMessage = 'Bad credentials.';
        const expectedState = { ...INITIAL_STATE, auth: { ...INITIAL_STATE.auth, errorMessage } };
        fetch.mockResponse(JSON.stringify({ errorMessage }), { status: 400 });

        store.dispatch(login('username', 'password'));

        store.subscribe(() => {
            expect(fetch.mock.calls[0][0]).toEqual(`${config.API_URL}login`);
            expect(store.getState()).toEqual(expectedState);
            expect(getAccessToken()).toBeNull();
            done();
        });
    });
});

/**
 * @param {function} submitCallback - Submit callback
 * @param {string} username - Username
 * @param {string} password - Password
 */
const expectedCallbackOnFormSubmitCalledOnceWith = (submitCallback, username, password) => {
    expect(submitCallback.mock.calls.length).toBe(1);
    expect(submitCallback.mock.calls[0][0]).toEqual(username);
    expect(submitCallback.mock.calls[0][1]).toEqual(password);
};

/**
 * @param {React.Component} component - Component
 */
const clickSubmitButton = (component) => {
    const node = ReactDOM.findDOMNode(component);
    const button = node.querySelector('button[type="submit"]');
    ReactTestUtils.Simulate.click(button);
};

/**
 * @param {React.Component} component - Component
 * @param {string} username - Username
 */
const fillUpUsernameField = (component, username) => {
    const node = ReactDOM.findDOMNode(component);
    const field = node.querySelector('#username');

    field.value = username;
    ReactTestUtils.Simulate.change(field);
};

/**
 * @param {React.Component} component - Component
 * @param {string} password - Password
 */
const fillUpPasswordField = (component, password) => {
    const node = ReactDOM.findDOMNode(component);
    const field = node.querySelector('input[type="password"]');

    field.value = password;
    ReactTestUtils.Simulate.change(field);
};
