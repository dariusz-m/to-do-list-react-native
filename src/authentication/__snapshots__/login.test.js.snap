// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth test Login component renders correctly with error message 1`] = `
<View
  style={
    Object {
      "textAlign": "center",
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 40,
        "fontWeight": "bold",
      }
    }
  >
    Sign In
  </Text>
  <View>
    <Text
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Username: 
    </Text>
    <TextInput
      allowFontScaling={true}
      autoCorrect={true}
      onChangeText={[Function]}
      placeholder="username..."
      secureTextEntry={false}
      style={
        Object {
          "borderColor": "black",
          "borderWidth": 0.2,
          "height": 40,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View>
    <Text
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Password: 
    </Text>
    <TextInput
      allowFontScaling={true}
      autoCorrect={true}
      onChangeText={[Function]}
      placeholder="password..."
      secureTextEntry={true}
      style={
        Object {
          "borderColor": "black",
          "borderWidth": 0.2,
          "height": 40,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    accessibilityComponentType="button"
    accessibilityTraits={
      Array [
        "button",
      ]
    }
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
            Object {
              "color": "blue",
            },
          ]
        }
      >
        Sign in
      </Text>
    </View>
  </View>
  <Text
    style={
      Object {
        "color": "red",
        "textAlign": "center",
      }
    }
  >
    <Text>
      Login failed!
      

    </Text>
    <Text>
      Bad credentials
    </Text>
  </Text>
</View>
`;

exports[`Auth test Login component renders correctly without error message 1`] = `
<View
  style={
    Object {
      "textAlign": "center",
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 40,
        "fontWeight": "bold",
      }
    }
  >
    Sign In
  </Text>
  <View>
    <Text
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Username: 
    </Text>
    <TextInput
      allowFontScaling={true}
      autoCorrect={true}
      onChangeText={[Function]}
      placeholder="username..."
      secureTextEntry={false}
      style={
        Object {
          "borderColor": "black",
          "borderWidth": 0.2,
          "height": 40,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View>
    <Text
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Password: 
    </Text>
    <TextInput
      allowFontScaling={true}
      autoCorrect={true}
      onChangeText={[Function]}
      placeholder="password..."
      secureTextEntry={true}
      style={
        Object {
          "borderColor": "black",
          "borderWidth": 0.2,
          "height": 40,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    accessibilityComponentType="button"
    accessibilityTraits={
      Array [
        "button",
      ]
    }
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
            Object {
              "color": "blue",
            },
          ]
        }
      >
        Sign in
      </Text>
    </View>
  </View>
</View>
`;
